compiler/src/dotty/tools/dotc/core/Symbols.scala

```diff
@@ -625,30 +622,59 @@ object Symbols {
 
     type ThisName = TypeName
 
+    type TreeOrProvider = AnyRef /* tpd.TreeProvider | tpd.PackageDef | tpd.TypeDef | tpd.EmptyTree | Null */
+
+    private[this] var myTree: TreeOrProvider = tpd.EmptyTree
+
     /** If this is either:
       *   - a top-level class and `-Yretain-trees` is set
      *    - a top-level class loaded from TASTY and `-tasty` or `-Xlink` is set
       * then return the TypeDef tree (possibly wrapped inside PackageDefs) for this class, otherwise EmptyTree.
       * This will force the info of the class.
       */
-    def tree(implicit ctx: Context): tpd.Tree /* tpd.PackageDef | tpd.TypeDef | tpd.EmptyTree */ = {
-      denot.info
-      // TODO: Consider storing this tree like we store lazy trees for inline functions
-      if (unpickler != null && !denot.isAbsent) {
-        assert(myTree.isEmpty)
-        val body = unpickler.body(ctx.addMode(Mode.ReadPositions))
-        myTree = body.headOption.getOrElse(tpd.EmptyTree)
-        if (!ctx.settings.fromTasty.value)
-          unpickler = null
-      }
-      myTree
+    def tree(implicit ctx: Context): Tree = treeContaining("")
+
+    /** Same as `tree` but load tree only if `id == ""` or the tree might contain `id`.
+     *  For Tasty trees this means consulting whether the name table defines `id`.
+     *  For already loaded trees, we maintain the referenced ids in an attachment.
+     */
+    def treeContaining(id: String)(implicit ctx: Context): Tree = denot.infoOrCompleter match {
+      case _: NoCompleter =>
+        tpd.EmptyTree
+      case _ =>
+        denot.ensureCompleted()
+        myTree match {
+          case fn: TreeProvider =>
+            if (id.isEmpty || fn.mightContain(id)) {
+              val tree = fn.tree
+              myTree = tree
+              tree
+            }
+            else tpd.EmptyTree
+          case tree: Tree @ unchecked =>
+            if (id.isEmpty || mightContain(tree, id)) tree else tpd.EmptyTree
+        }
     }
-    private[this] var myTree: tpd.Tree /* tpd.PackageDef | tpd.TypeDef | tpd.EmptyTree */ = tpd.EmptyTree
-    private[dotc] var unpickler: tasty.DottyUnpickler = _
 
-    private[dotc] def registerTree(tree: tpd.TypeDef)(implicit ctx: Context): Unit = {
-      if (ctx.settings.YretainTrees.value)
-        myTree = tree
+    def treeOrProvider: TreeOrProvider = myTree
+
+    private[dotc] def treeOrProvider_=(t: TreeOrProvider)(implicit ctx: Context): Unit =
+      myTree = t
+
+    private def mightContain(tree: Tree, id: String)(implicit ctx: Context): Boolean = {
+      val ids = tree.getAttachment(Ids) match {
+        case Some(ids) => ids
+        case None =>
+          val idSet = mutable.SortedSet[String]()
+          tree.foreachSubTree {
+            case tree: tpd.NameTree => idSet += tree.name.toString
```

Comment:
Should be restricted to `SimpleName`, like in `DottyUnpickler#mightContain` I think.